### PR TITLE
Successive upgrades in hotbuild via keybind

### PR DESCRIPTION
**Pull request by ChessBerry**
Hello everyone, 

this PR is a proposed feature addition to hotbuild, namely being able to queue successive upgrades using the upgrade keybind, instead of having to use a mouse click.

Upgrading e.g. a T1 factory to T3 in one go previously required a mouse click to queue the upgrade to T3, if the upgrade to T2 was not finished. 
Now, pressing the upgrade hotkey on your keyboard a second time does queue the upgrade to T3, regardless of if the T2 upgrade is finished or not.

I tested it with factories, radars and sonars of different factions, and even with adding units into a factory queue between the T2 and T3 upgrade and it seems to work as intended.

TODO: I don't know how to adjust the icon in the command card so that it shows the correct keybind as, right now, the T3 upgrade doesn't show a keybind at all if the T2 upgrade is queued but not finished yet.

Tell me what you think :)

### CODE REVIEW

**Comment by Garanas** [comment]
Looks like a good quality of life change. What is the behavior that you describe in the TODO? I don't entirely understand what it means / what you're trying to achieve. A screenshot can help.

**Comment by ChessBerry** [comment]
Let's say I have hotbuild-upgrade bound to M and I select a T1 factory. I will see the following in my command card (or whatever it's called):
Forged Alliance 14_09_2022 14_23_30 (2)
(Ignore the keybinds for all the other units)

If I now queue the upgrade to a T2 HQ, I will get access to the T2 build suite for unit queueing purposes, but notice that there is no keybind for the upgrade to the T3 HQ on the T3 HQ icon:
Forged Alliance 14_09_2022 14_23_40 (2)

After the upgrade to T2 is done, the T3 icon _does_ get the appropriate keybind:
Forged Alliance 14_09_2022 14_23_52 (2)

What is TODO is to give the icon in the intermediate step, when the T2 upgrade is queued but not finished, where I drew the red arrows, to correct keybind symbol, in this case the letter M. 
I don't know how to do that do as I haven't been able to find the appropriate place in the code governing these icon changes

**Comment by 4z0t** [comment]
Afaik it looks what blueprints it can build and upgrade into in construction.lua (have a nice day looking through __this__ code...)

**Comment by ChessBerry** [comment]
Took me a while, but I found where the hotbuild upgrade icons are set in gamemain.lua. The icons and successive upgrades should now work correctly, even for factories or unconventional structures, like the four times upgradeable cybran shield.

**Comment by Garanas** [comment]
Exciting, not sure why that code is there. I'll look at this tonight, see if I can break it 😄

**Comment by ChessBerry** [comment]
I phrased that a bit misleadingly: The icons themselves are not directly set in gamemain.lua, but the table that governs what is a valid upgrade target (according to hotbuild) is.

Also yes, please break it, there is for sure stuff I've overlooked that could be improved :P

**Comment by Garanas** [comment]
It doesn't work for build mode, but I think I'm the only user of that. @Hdt80bro do you happen to use hotbuild, and if so can you review this?

**Comment by Hdt80bro** [comment]
Sorry, I don't

**Comment by Garanas** [comment]
@4z0t do you use hotbuild, if so can you review this?

**Comment by 4z0t** [comment]
Yes, It works! Nice job!

File: lua/keymap/hotbuild.lua [hunk 1]
```diff
@@ -571,32 +571,50 @@ function buildActionUnit(name, modifier)
...
+-- Recursively finds and tries to order possible upgrades from highest tech to lowest tech. This allows queuing upgrades
+-- successively using keybinds, even if the previous upgrade isn't finished yet.
+-- Doing this fully recursively is probably overkill as the support fac / HQ is the only time when there are more than
+-- two possible upgrade paths, and the cybran shield is the only building with more than two successive upgrades, but
+-- hacking it together didn't feel right, hence the 'better' recursive solution.
+function issueUpgradeCommand(upgrades, buildableCategories)
+    local success = false
+    for _, upgr in upgrades do
+        local bp = __blueprints[upgr]
+        successiveUpgrades = upgradeTab[bp.BlueprintId] or {bp.General.UpgradesTo}
```
Comment (4z0t) [inline lua/keymap/hotbuild.lua:584]: Did you want it to be `local` variable?

File: lua/keymap/hotbuild.lua [hunk 2]
```diff
@@ -571,32 +571,50 @@ function buildActionUnit(name, modifier)
...
 end
 
--- Does not upgrade T1 facs that are currently upgrading to T2 to T3 when issued
+-- Helper function for the buildActionUpgrade() function below. 
+-- Recursively finds and tries to order possible upgrades from highest tech to lowest tech. This allows queuing upgrades
+-- successively using keybinds, even if the previous upgrade isn't finished yet.
+-- Doing this fully recursively is probably overkill as the support fac / HQ is the only time when there are more than
+-- two possible upgrade paths, and the cybran shield is the only building with more than two successive upgrades, but
+-- hacking it together didn't feel right, hence the 'better' recursive solution.
+function issueUpgradeCommand(upgrades, buildableCategories)
```
Comment (4z0t) [inline lua/keymap/hotbuild.lua:580]: PascalCase

File: lua/keymap/hotbuild.lua [hunk 3]
```diff
@@ -571,32 +571,50 @@ function buildActionUnit(name, modifier)
...
+    local success = false
+    for _, upgr in upgrades do
```
Comment (4z0t) [inline lua/keymap/hotbuild.lua:582]: Dont short names

**Review by 4z0t — CHANGES_REQUESTED** [review]

File: lua/keymap/hotbuild.lua (hunk 3 above)
Comment (ChessBerry) [reply lua/keymap/hotbuild.lua:582]: So,
```
for _, u in upgrades do
```
would be better?

File: lua/keymap/hotbuild.lua (hunk 1 above)
Comment (ChessBerry) [reply lua/keymap/hotbuild.lua:584]: yes, there is no reason for successiveUpgrades to be global

File: lua/keymap/hotbuild.lua (hunk 2 above)
Comment (ChessBerry) [reply lua/keymap/hotbuild.lua:580]: PascalCase for the issueUpgradeCommand function? All other functions in hotbuild.lua are camelCase though. Is the new, proper style guide to make all new functions PascalCase?

Comment (Garanas) [reply lua/keymap/hotbuild.lua:580]: Yes - we're trying to get a standard going. New content should use pascal case.

File: lua/keymap/hotbuild.lua (hunk 3 above)
Comment (Hdt80bro) [reply lua/keymap/hotbuild.lua:582]: It's a matter of style--the specific tradition 4z0t is referring to is one that uses one-letter variable names for loop iterators.  That's fine when you don't have lots of nested loops like you have here, but when there are--or when the variable does lots of heavy lifting--I prefer to see it spelled out completely (unless that creates a variable that has only a one-letter *difference* from another variable [e.g. `upgrade` and `upgrades`]). Abbreviations in general are fine. We don't actually a unifying standard on variable names, so pretty much whatever you do (within reason) is fine.

Comment (ChessBerry) [reply lua/keymap/hotbuild.lua]: I changed 'upgr' to 'u' as mentioned above

**Review by Garanas — APPROVED** [review]

**Comment by ChessBerry** [comment]
Anything still missing that's preventing a merge? 

If I understand correctly, I implemented the last requested change in 7a4c18c above

**Review by 4z0t — APPROVED** [review]

**Comment by Garanas** [comment]
Yeah, I was discussing this in Zulip. Github doesn't understand when the requested changes are applied, hence we need 4z0t to do another review.

**Pull request merged**